lib/Sema/CSSolver.cpp

```diff
@@ -1860,26 +1860,29 @@ static Constraint *selectBestBindingDisjunction(
   // type variable.
   SmallVector<Constraint *, 8> bindingDisjunctions;
   for (auto *disjunction : disjunctions) {
-    llvm::Optional<TypeVariableType *> commonTypeVariable;
+    TypeVariableType *commonTypeVariable = nullptr;
     if (llvm::all_of(
             disjunction->getNestedConstraints(),
             [&](Constraint *bindingConstraint) {
               if (bindingConstraint->getKind() != ConstraintKind::Bind)
                 return false;
 
-              auto *tv =
-                  bindingConstraint->getFirstType()->getAs<TypeVariableType>();
+              auto *tv = cs.simplifyType(bindingConstraint->getFirstType())
```

Comment:
Maybe a better choice in this case would be to skip disjunction completely if it already has its type variable bound to something?

Reply:
In the example you've added force unwrap disjunction is bound to ` @lvalue ImplicitlyUnwrapped.x: $T1 == @lvalue Int?` by the time we get to `selectBestBindingDisjunction`.

Reply:
Well, I don't know if the case where the original first type has been merged with some other type variable is important to handle here.  It at least seems plausible.

Reply:
If we have simplified type variable to something else once why do it for every choice? That disqualifies disjunction right away...

Reply:
It's not necessarily simplified to a concrete type.  It is in my example, but I don't think it would have to be.

I'm just trying to stop this from crashing on actual code.  If you think there's a significant improvement to the algorithm possible, I'd rather leave that to you.

Honestly, the whole structure of this function seems strange to me: it's building up a list of viable disjunctions, then doing a second pass to preferentially pick one if it satisfies some obscure extra condition, then if that doesn't work just using the first viable one.  Why not check the extra condition in the first pass and fast-path out immediately if it passes rather than considering all the other disjunctions first, and just save the first viable disjunction in case we don't ever find one that passes the extra condition?

Reply:
+1, I think this could be made better. Disjunction should either have a single type variable on the or function type e.g. `($T0) -> $T1` for initializers on left-hand side of every choice, which means that all of the choices should be simplified to the same type (on left-hand side), I'm not sure why it's checked in the loop.

@rudkx could you please shed some light on this?

Reply:
No worries, I will handle it!